runner/src/runner.rs

```diff
@@ -100,23 +100,206 @@ impl Runner {
+        Ok((bin_name.to_string(), bin_path))
+    }
+
+    fn delete_downloaded_release(runner: &mut impl RunnerExt) -> Result<(), Error> {
+        let release = runner.downloaded_release().as_ref().ok_or(Error::NoDownloadedRelease)?;
+        log::info!("Removing old release, with path {:?}", release.path);
+
+        retry_with_log(
+            || Ok(fs::remove_file(&release.path)?),
+            "Failed to remove old release".to_string(),
+        )?;
+
+        runner.set_downloaded_release(None);
+        Ok(())
+    }
+
+    fn terminate_proc_and_wait(runner: &mut impl RunnerExt) -> Result<u32, Error> {
+        let child_proc = runner.child_proc().as_mut().ok_or(Error::NoChildProcess)?;
+
+        let _ = retry_with_log(
+            || {
+                Ok(signal::kill(
+                    Pid::from_raw(child_proc.id().try_into().map_err(|_| Error::IntegerConversionError)?),
+                    Signal::SIGTERM,
+                ))
+            },
+            "Failed to kill child process".to_string(),
+        )?;
+
+        match child_proc.wait() {
+            Ok(exit_status) => log::info!(
+                "Terminated vault process (pid: {}) with exit status {}",
+                child_proc.id(),
+                exit_status
+            ),
+            Err(error) => log::warn!("Vault process termination error: {}", error),
+        };
+        let pid = child_proc.id();
+        runner.set_child_proc(None);
+        Ok(pid)
+    }
+
+    async fn try_get_release<T: RunnerExt + StorageReader>(
+        runner: &T,
+        pending: bool,
+    ) -> Result<Option<ClientRelease>, Error> {
+        let storage_item = if pending {
+            PENDING_RELEASE_STORAGE_ITEM
+        } else {
+            CURRENT_RELEASE_STORAGE_ITEM
+        };
+
+        let storage_key = runner.compute_storage_key(PARACHAIN_MODULE.to_string(), storage_item.to_string());
+
+        retry_with_log_async(
+            || {
+                let key = &storage_key;
+                runner
+                    .read_chain_storage::<ClientRelease>(runner.parachain_rpc(), Some(key.to_string()))
+                    .into_future()
+                    .boxed()
+            },
+            "Error fetching executable".to_string(),
+        )
+        .await
+    }
+
+    async fn read_chain_storage<T: 'static + Decode + Debug, V: RunnerExt + StorageReader>(
+        runner: &V,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+    ) -> Result<Option<T>, Error> {
+        let enc_res = runner
+            .query_storage(parachain_rpc, maybe_storage_key, "state_getStorage".to_string())
+            .await;
+        enc_res
+            .map(|r| {
+                let v = r.to_vec();
+                T::decode(&mut &v[..])
+            })
+            .transpose()
+            .map_err(Into::into)
+    }
+
+    pub async fn run(mut runner: Box<dyn RunnerExt + Send>, mut shutdown_signals: SignalsInfo) -> Result<(), Error> {
```

Comment:
To follow up on a point from a previous PR @savudani8 did you consider adding this as a [default implementation](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations) on `RunnerExt`?

Reply:
The `run` method requires a boxed reference to the trait object for mocking purposes, and since trait fns can't take full ownership of `self`, I don't know how that could be done.